File: src/main/clients/gradescope-client.ts
import * as cheerio from 'cheerio';
import { BrowserWindow } from 'electron';
import { loadCookies, cookieHeader, type StoredCookie } from '../auth/cookie-store';
import { authenticateCalNet } from '../auth/calnet-auth';
import type { Assignment } from '../../shared/types/assignment';

const GRADESCOPE_URL = 'https://www.gradescope.com';
const LOOKAHEAD_HOURS = 168;

function getCredentials(): { calnetId: string; calnetPassphrase: string } {
  const email = process.env.GRADESCOPE_EMAIL || '';
  const password = process.env.GRADESCOPE_PASSWORD || '';
  return {
    calnetId: email.split('@')[0],
    calnetPassphrase: password,
  };
}

async function fetchWithCookies(url: string, cookies: StoredCookie[]): Promise<Response> {
  return fetch(url, {
    headers: {
      Cookie: cookieHeader(cookies),
    },
    redirect: 'manual',
  });
}

export async function cookiesValid(): Promise<boolean> {
  const cookies = loadCookies('gradescope');
  if (!cookies) return false;

  try {
    const resp = await fetchWithCookies(`${GRADESCOPE_URL}/account`, cookies);
    return resp.status === 200;
  } catch {
    return false;
  }
}

export async function loginViaCalNet(): Promise<boolean> {
  const { calnetId, calnetPassphrase } = getCredentials();

  try {
    const cookies = await authenticateCalNet({
      url: `${GRADESCOPE_URL}/saml`,
      cookieName: 'gradescope',
      successUrlPattern: 'gradescope.com',
      calnetId,
      calnetPassphrase,
      preCalnetSteps: async (win: BrowserWindow) => {
        // Wait for the SAML school-selection page to fully load
        await new Promise((r) => setTimeout(r, 2000));

        // Type "Berkeley" using keyboard events (not .value) so React's
        // search input actually filters. Then find and click the CalNet result.
        await win.webContents.executeJavaScript(`
          (function() {
            const input = document.querySelector("input[type='text'], input[placeholder]");
            if (!input) return;
            input.focus();
            // Use native setter + InputEvent to trigger React state update
            const nativeSetter = Object.getOwnPropertyDescriptor(
              window.HTMLInputElement.prototype, 'value'
            ).set;
            nativeSetter.call(input, 'Berkeley');
            input.dispatchEvent(new Event('input', { bubbles: true }));
          })();
        `);

        // Wait for search results to filter
        await new Promise((r) => setTimeout(r, 2000));

        // Find and click the CalNet/Berkeley result specifically
        await win.webContents.executeJavaScript(`
          (function() {
            const items = document.querySelectorAll('div.samlProvider--name');
            for (const item of items) {
              const text = item.textContent || '';
              if (text.includes('Berkeley') || text.includes('CalNet')) {
                item.click();
                return;
              }
            }
            // Fallback: click the first result if only one showed up
            if (items.length === 1) items[0].click();
          })();
        `);
      },
    });
    return cookies.length > 0;
  } catch (err) {
    console.error('[Gradescope] Login failed:', err);
    return false;
  }
}

function parseDate(text: string): Date | null {
  let cleaned = text.replace(/\n/g, ' ').trim();

  for (const prefix of ['Due ', 'Due: ', 'due ', 'Due Date: ']) {
    if (cleaned.startsWith(prefix)) {
      cleaned = cleaned.slice(prefix.length);
    }
  }
  cleaned = cleaned.trim();

  // Try various Gradescope date formats
  // "Mar 15, 2026 11:59 PM", "Mar 15, 2026 at 11:59 PM", "Mar 15 at 11:59 PM"
  const patterns = [
    /^(\w{3,9})\s+(\d{1,2}),?\s+(\d{4})\s+(?:at\s+)?(\d{1,2}):(\d{2})\s*(AM|PM)$/i,
    /^(\w{3,9})\s+(\d{1,2})\s+(?:at\s+)?(\d{1,2}):(\d{2})\s*(AM|PM)$/i,
  ];

  const months: Record<string, number> = {
    jan: 0, feb: 1, mar: 2, apr: 3, may: 4, jun: 5,
    jul: 6, aug: 7, sep: 8, oct: 9, nov: 10, dec: 11,
    january: 0, february: 1, march: 2, april: 3, june: 5,
    july: 6, august: 7, september: 8, october: 9, november: 10, december: 11,
  };

  // Pattern 1: Full date with year
  const m1 = cleaned.match(patterns[0]);
  if (m1) {
    const month = months[m1[1].toLowerCase()];
    if (month === undefined) return null;
    const day = parseInt(m1[2]);
    const year = parseInt(m1[3]);
    let hour = parseInt(m1[4]);
    const minute = parseInt(m1[5]);
    const ampm = m1[6].toUpperCase();
    if (ampm === 'PM' && hour !== 12) hour += 12;
    if (ampm === 'AM' && hour === 12) hour = 0;

    // Gradescope displays in Pacific time
    const pacific = new Date(Date.UTC(year, month, day, hour + 8, minute)); // Approximate PST=UTC-8
    return pacific;
  }

  // Pattern 2: Date without year
  const m2 = cleaned.match(patterns[1]);
  if (m2) {
    const month = months[m2[1].toLowerCase()];
    if (month === undefined) return null;
    const day = parseInt(m2[2]);
    const year = new Date().getFullYear();
    let hour = parseInt(m2[3]);
    const minute = parseInt(m2[4]);
    const ampm = m2[5].toUpperCase();
    if (ampm === 'PM' && hour !== 12) hour += 12;
    if (ampm === 'AM' && hour === 12) hour = 0;

    const pacific = new Date(Date.UTC(year, month, day, hour + 8, minute));
    return pacific;
  }

  return null;
}

export async function getUpcomingAssignments(): Promise<Assignment[]> {
  const cookies = loadCookies('gradescope');
  if (!cookies) return [];

  const now = new Date();
  const cutoff = new Date(now.getTime() + LOOKAHEAD_HOURS * 60 * 60 * 1000);
  const assignments: Assignment[] = [];

  try {
    // Fetch account page for course list
    const accountResp = await fetchWithCookies(`${GRADESCOPE_URL}/account`, cookies);
    if (accountResp.status !== 200) return assignments;

    const html = await accountResp.text();
    const $ = cheerio.load(html);

    // Find course links
    const seenCourses = new Set<string>();
    const courseLinks: Array<{ id: string; name: string }> = [];

    $('a[href*="/courses/"]').each((_i, el) => {
      const href = $(el).attr('href') || '';
      const match = href.match(/\/courses\/(\d+)/);
      if (!match) return;
      const courseId = match[1];
      if (seenCourses.has(courseId)) return;
      seenCourses.add(courseId);

      const shortName = $(el).find('h3.courseBox--shortname').text().trim();
      const fullName = $(el).find('div.courseBox--name').text().trim();
      const courseName = shortName || fullName || $(el).text().trim() || `Course ${courseId}`;

      courseLinks.push({ id: courseId, name: courseName });
    });

    // Fetch assignments for each course
    for (const course of courseLinks) {
      try {
        const courseResp = await fetchWithCookies(
          `${GRADESCOPE_URL}/courses/${course.id}`,
          cookies,
        );
        if (courseResp.status !== 200) continue;

        const courseHtml = await courseResp.text();
        const $c = cheerio.load(courseHtml);

        const rows = $c('tr[role="row"]').length > 0
          ? $c('tr[role="row"]')
          : $c('tr');

        rows.each((_i, row) => {
          const cells = $c(row).find('td');
          if (cells.length === 0) return;

          const th = $c(row).find('th');
          const name = th.length > 0
            ? th.text().trim()
            : $c(cells[0]).text().trim();

          // Get assignment ID
          let assignId: string | null = null;
          const assignLink = $c(row).find('a[href*="/assignments/"]');
          if (assignLink.length > 0) {
            const href = assignLink.attr('href') || '';
            const match = href.match(/\/assignments\/(\d+)/);
            if (match) assignId = match[1];
          }
          if (!assignId) {
            const submitBtn = $c(row).find('button.js-submitAssignment');
            if (submitBtn.length > 0) {
              assignId = submitBtn.attr('data-assignment-id') || null;
            }
          }
          if (!assignId) return;

          // Parse due date
          let dueAt: Date | null = null;
          const dueEl = $c(row).find('.submissionTimeChart--dueDate');
          if (dueEl.length > 0 && dueEl.attr('datetime')) {
            try {
              dueAt = new Date(dueEl.attr('datetime')!);
            } catch { /* ignore */ }
          }

          if (!dueAt) {
            cells.each((_j, cell) => {
              if (dueAt) return;
              const parsed = parseDate($c(cell).text().trim());
              if (parsed) dueAt = parsed;
            });
          }

          if (!dueAt) return;
          if (dueAt <= now || dueAt > cutoff) return;

          // Check submission status
          const rowText = $c(row).text();
          const submitted =
            rowText.includes('Submitted') ||
            $c(row).find('.submissionStatus--submitted').length > 0 ||
            $c(row).find('.workflowCheck--complete').length > 0;

          assignments.push({
            id: `gs_${course.id}_${assignId}`,
            name,
            courseName: course.name,
            dueAt: dueAt.toISOString(),
            url: `${GRADESCOPE_URL}/courses/${course.id}/assignments/${assignId}`,
            source: 'gradescope',
            submitted,
          });
        });
      } catch (err) {
        console.error(`[Gradescope] Error fetching assignments for ${course.name}:`, err);
      }
    }
  } catch (err) {
    console.error('[Gradescope] Error:', err);
  }

  return assignments;
}

File: legacy/gradescope_client.py
import json
import os
import time
from datetime import datetime, timezone, timedelta
from zoneinfo import ZoneInfo

import requests
from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

from config import CALNET_ID, CALNET_PASSPHRASE, LOOKAHEAD_HOURS

COOKIES_PATH = os.path.join(os.path.dirname(__file__), "gradescope_cookies.json")
GRADESCOPE_URL = "https://www.gradescope.com"
PACIFIC = ZoneInfo("America/Los_Angeles")


def _save_cookies(cookies):
    with open(COOKIES_PATH, "w") as f:
        json.dump(cookies, f)


def _load_cookies():
    if os.path.exists(COOKIES_PATH):
        with open(COOKIES_PATH) as f:
            return json.load(f)
    return None


def _apply_cookies(session, cookies):
    for cookie in cookies:
        session.cookies.set(cookie["name"], cookie["value"], domain=cookie.get("domain", ""))


def _cookies_valid(cookies):
    """Check if saved cookies still give us an authenticated session."""
    session = requests.Session()
    _apply_cookies(session, cookies)
    resp = session.get(f"{GRADESCOPE_URL}/account", allow_redirects=False)
    return resp.status_code == 200


def _login_via_calnet():
    """Open Edge browser for CalNet SSO login, return cookies after auth."""
    options = webdriver.EdgeOptions()
    driver = webdriver.Edge(options=options)

    try:
        # Go to Gradescope's SAML school-selection page
        driver.get(f"{GRADESCOPE_URL}/saml")

        # Wait for the page to load, then search for and select "CalNet ID"
        print("  [Gradescope] Selecting CalNet ID on SAML page...", flush=True)
        WebDriverWait(driver, 15).until(
            EC.presence_of_element_located((By.CSS_SELECTOR, "input[type='text'], input[placeholder]"))
        )
        search_input = driver.find_element(By.CSS_SELECTOR, "input[type='text'], input[placeholder]")
        search_input.send_keys("CalNet")
        time.sleep(1.5)

        # Click the "CalNet ID" result from the dropdown
        calnet_btn = WebDriverWait(driver, 10).until(
            EC.element_to_be_clickable((By.CSS_SELECTOR, "div.samlProvider--name"))
        )
        calnet_btn.click()

        # Wait for CalNet login page to load (may take a moment for redirect)
        print("  [Gradescope] Waiting for CalNet login page...", flush=True)
        WebDriverWait(driver, 30).until(EC.url_contains("auth.berkeley.edu"))
        WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.ID, "username"))
        )

        # Auto-fill credentials
        driver.find_element(By.ID, "username").send_keys(CALNET_ID)
        driver.find_element(By.ID, "password").send_keys(CALNET_PASSPHRASE)
        driver.find_element(By.ID, "submitBtn").click()

        # Wait for redirect back to Gradescope — user may need to handle Duo 2FA
        print("  [Gradescope] Waiting for CalNet auth (complete Duo if prompted)...", flush=True)
        WebDriverWait(driver, 180).until(
            EC.url_contains("gradescope.com")
        )

        time.sleep(2)
        cookies = driver.get_cookies()
        _save_cookies(cookies)
        print("  [Gradescope] Login successful, cookies saved.", flush=True)
        return cookies
    finally:
        driver.quit()


def _get_session():
    """Get an authenticated requests session for Gradescope."""
    cookies = _load_cookies()
    if cookies and _cookies_valid(cookies):
        session = requests.Session()
        _apply_cookies(session, cookies)
        return session

    # Cookies missing or expired — need fresh login
    cookies = _login_via_calnet()
    session = requests.Session()
    _apply_cookies(session, cookies)
    return session


def _parse_date(text):
    """Parse Gradescope's displayed date into a UTC datetime."""
    text = text.replace("\n", " ").strip()
    for prefix in ("Due ", "Due: ", "due ", "Due Date: "):
        if text.startswith(prefix):
            text = text[len(prefix):]

    formats = [
        "%b %d, %Y %I:%M %p",
        "%b %d, %Y at %I:%M %p",
        "%B %d, %Y %I:%M %p",
        "%B %d, %Y at %I:%M %p",
        "%b %d at %I:%M %p",
    ]
    for fmt in formats:
        try:
            dt = datetime.strptime(text.strip(), fmt)
            # If year is 1900 (missing from format), assume current year
            if dt.year == 1900:
                dt = dt.replace(year=datetime.now().year)
            return dt.replace(tzinfo=PACIFIC).astimezone(timezone.utc)
        except ValueError:
            continue
    return None


def cookies_valid() -> bool:
    """Check whether saved Gradescope cookies are still valid."""
    cookies = _load_cookies()
    if not cookies:
        return False
    return _cookies_valid(cookies)


def login_via_calnet() -> bool:
    """Run the CalNet/Duo login flow and return True on success."""
    try:
        cookies = _login_via_calnet()
        return cookies is not None and len(cookies) > 0
    except Exception as e:
        print(f"  [Gradescope] Login failed: {e}", flush=True)
        return False


def get_upcoming_assignments() -> list[dict]:
    """Fetch assignments due within the lookahead window from Gradescope."""
    session = _get_session()
    now = datetime.now(timezone.utc)
    cutoff = now + timedelta(hours=LOOKAHEAD_HOURS)
    assignments = []

    # Fetch account page to get course list (not the home page)
    resp = session.get(f"{GRADESCOPE_URL}/account")
    soup = BeautifulSoup(resp.text, "html.parser")

    # Courses live inside div.courseList as <a> tags with /courses/{id} hrefs
    course_links = soup.select("div.courseList a[href*='/courses/']")
    if not course_links:
        # Broader fallback
        course_links = soup.select("a[href*='/courses/']")

    seen_courses = set()
    for link in course_links:
        href = link.get("href", "")
        if "/courses/" not in href:
            continue
        parts = href.strip("/").split("/")
        try:
            idx = parts.index("courses")
            course_id = parts[idx + 1]
        except (ValueError, IndexError):
            continue
        if course_id in seen_courses:
            continue
        seen_courses.add(course_id)

        # Extract course name from courseBox child elements
        short = link.select_one("h3.courseBox--shortname")
        full = link.select_one("div.courseBox--name")
        course_name = (
            (short.get_text(strip=True) if short else None)
            or (full.get_text(strip=True) if full else None)
            or link.get_text(strip=True)
            or f"Course {course_id}"
        )

        try:
            # Student view: /courses/{id} shows the assignment table
            resp = session.get(f"{GRADESCOPE_URL}/courses/{course_id}")
            asoup = BeautifulSoup(resp.text, "html.parser")

            # Assignment rows use role="row" in the student view table
            rows = asoup.select("tr[role='row']")
            if not rows:
                rows = asoup.select("tr")

            for row in rows:
                cells = row.select("td")
                if not cells:
                    continue

                # Assignment name lives in the row's <th>, not <td>
                th = row.select_one("th")
                name = th.get_text(strip=True) if th else cells[0].get_text(strip=True)

                # Assignment ID: from a link or a submit button
                assign_id = None
                assign_link = row.select_one("a[href*='/assignments/']")
                if assign_link:
                    assign_id = assign_link.get("href", "").strip("/").split("/")[-1]
                else:
                    submit_btn = row.select_one("button.js-submitAssignment")
                    if submit_btn:
                        assign_id = submit_btn.get("data-assignment-id")
                    else:
                        continue

                if not assign_id:
                    continue

                # Due date: prefer datetime attribute on submissionTimeChart--dueDate
                due_at = None
                due_el = row.select_one(".submissionTimeChart--dueDate")
                if due_el and due_el.get("datetime"):
                    try:
                        due_at = datetime.fromisoformat(
                            due_el["datetime"].replace("Z", "+00:00")
                        )
                    except ValueError:
                        pass

                # Fallback: parse displayed text in cells
                if not due_at:
                    for cell in cells:
                        parsed = _parse_date(cell.get_text(strip=True))
                        if parsed:
                            due_at = parsed
                            break

                if not due_at:
                    continue

                if now < due_at <= cutoff:
                    # Check submission status
                    row_text = row.get_text()
                    submitted = (
                        "Submitted" in row_text
                        or row.select_one(".submissionStatus--submitted") is not None
                        or row.select_one(".workflowCheck--complete") is not None
                    )

                    assignments.append({
                        "id": f"gs_{course_id}_{assign_id}",
                        "name": name,
                        "course_name": course_name,
                        "due_at": due_at,
                        "url": f"{GRADESCOPE_URL}/courses/{course_id}/assignments/{assign_id}",
                        "source": "gradescope",
                        "submitted": submitted,
                    })
        except Exception as e:
            print(f"  [Gradescope] Error fetching assignments for {course_name}: {e}", flush=True)

    return assignments
